src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -523,7 +523,7 @@ private bool GetEnumeratorInfoAndInferCollectionElementType(ref ForEachEnumerato
         {
             UnwrapCollectionExpressionIfNullable(ref collectionExpr, diagnostics);
 
-            bool gotInfo = GetEnumeratorInfo(ref builder, collectionExpr, diagnostics);
+            bool gotInfo = GetEnumeratorInfo(ref builder, collectionExpr, isAsyncOverride: null, diagnostics);
```

Comment:
>gotInfo [](start = 17, length = 7)

:memo: The changes in this file are to try binding the enumerator info a second time in error cases, but assuming a different async-ness. Maybe you were using a `foreach` on a type that is `IAsyncEnumerable`, or vice-versa.